File: src/secrets.ts
import * as A from "@automerge/automerge";
import { randomUUIDv7 } from "bun";
import type { Secret, Workspace } from "./types";

export function addSecret(
  doc: A.Doc<Workspace>,
  fields: Omit<Secret, "id">
): A.Doc<Workspace> {
  const id = randomUUIDv7();
  return A.change(doc, `add secret ${fields.name}`, (d) => {
    d.secrets[id] = { id, ...fields };
  });
}

export function removeSecret(
  doc: A.Doc<Workspace>,
  id: string
): A.Doc<Workspace> {
  return A.change(doc, `remove secret ${id}`, (d) => {
    delete d.secrets[id];
    // Remove from any projects that reference it
    for (const project of Object.values(d.projects)) {
      const idx = project.secret_ids.indexOf(id);
      if (idx !== -1) project.secret_ids.splice(idx, 1);
    }
  });
}

export function updateSecret(
  doc: A.Doc<Workspace>,
  id: string,
  fields: Omit<Secret, "id">
): A.Doc<Workspace> {
  return A.change(doc, `update secret ${id}`, (d) => {
    const s = d.secrets[id];
    if (!s) throw new Error(`Secret ${id} not found`);
    s.name = fields.name;
    s.value = fields.value;
    s.description = fields.description;
    s.tags.splice(0, s.tags.length, ...fields.tags);
  });
}

export function listSecrets(doc: A.Doc<Workspace>): Secret[] {
  return Object.values(doc.secrets);
}

File: lib/src/secrets.rs
use crate::{
    crypto::{decrypt_field, encrypt_field},
    error::Result,
    vault_document::{PlaintextSecret, Secret, VaultDocument},
};
use automerge::AutoCommit;
use autosurgeon::{hydrate, reconcile};
use uuid::Uuid;

fn encrypt_secret(
    fields: &PlaintextSecretFields,
    dek: &[u8; 32],
) -> Result<(String, String, String, String)> {
    Ok((
        encrypt_field(&fields.name, dek)?,
        encrypt_field(&fields.value, dek)?,
        encrypt_field(&fields.description, dek)?,
        encrypt_field(&serde_json::to_string(&fields.tags)?, dek)?,
    ))
}

fn decrypt_secret(s: &Secret, dek: &[u8; 32]) -> Result<PlaintextSecret> {
    let tags_json = decrypt_field(&s.tags, dek)?;
    let tags: Vec<String> = serde_json::from_str(&tags_json)?;
    Ok(PlaintextSecret {
        id: s.id.clone(),
        name: decrypt_field(&s.name, dek)?,
        value: decrypt_field(&s.value, dek)?,
        description: decrypt_field(&s.description, dek)?,
        tags,
    })
}

pub struct PlaintextSecretFields {
    pub name: String,
    pub value: String,
    pub description: String,
    pub tags: Vec<String>,
}

pub fn add_secret(
    doc: &mut AutoCommit,
    dek: &[u8; 32],
    fields: PlaintextSecretFields,
) -> Result<()> {
    let id = Uuid::now_v7().to_string();
    let (enc_name, enc_value, enc_desc, enc_tags) = encrypt_secret(&fields, dek)?;

    let mut vault_doc: VaultDocument = hydrate(doc)?;
    vault_doc.secrets.insert(
        id.clone(),
        Secret {
            id,
            name: enc_name,
            value: enc_value,
            description: enc_desc,
            tags: enc_tags,
        },
    );
    reconcile(doc, &vault_doc)?;
    Ok(())
}

pub fn remove_secret(doc: &mut AutoCommit, id: &str) -> Result<()> {
    let mut vault_doc: VaultDocument = hydrate(doc)?;
    vault_doc.secrets.remove(id);
    reconcile(doc, &vault_doc)?;
    Ok(())
}

pub fn update_secret(
    doc: &mut AutoCommit,
    dek: &[u8; 32],
    id: &str,
    fields: PlaintextSecretFields,
) -> Result<()> {
    let (enc_name, enc_value, enc_desc, enc_tags) = encrypt_secret(&fields, dek)?;
    let mut vault_doc: VaultDocument = hydrate(doc)?;

    let secret = vault_doc
        .secrets
        .get_mut(id)
        .ok_or_else(|| crate::error::Error::SecretNotFound(id.to_string()))?;
    secret.name = enc_name;
    secret.value = enc_value;
    secret.description = enc_desc;
    secret.tags = enc_tags;

    reconcile(doc, &vault_doc)?;
    Ok(())
}

pub fn list_secrets(doc: &AutoCommit, dek: &[u8; 32]) -> Result<Vec<PlaintextSecret>> {
    let vault_doc: VaultDocument = hydrate(doc)?;
    vault_doc
        .secrets
        .values()
        .map(|s| decrypt_secret(s, dek))
        .collect()
}
